ships/0006-surface-results-in-buildruns.md

```diff
@@ -0,0 +1,293 @@
+
+### Build Results
+
+These are results defined at runtime by the BuildRun controller. In this proposal we categorize them in two:
+
+- sources
+- output
+
+Following what was mentioned in [EP removal of tekton resources](https://github.com/shipwright-io/build/blob/main/docs/proposals/removal-tekton-resources.md#3-tekton-results), we need to consider the following:
+
+#### Current State of Build Results for Sources
+
+Assuming the following Build definition:
+
+```yaml
+apiVersion: shipwright.io/v1alpha1
+kind: Build
+metadata:
+  name: a-build
+spec:
+  source:
+    url: https://github.com/shipwright-io/sample-go
+
+```
+
+We will have the following:
+
+| Result Name                       | Naming Convention                      | Emitted Today | Definition  | Notes                                |
+| --------------------------------: | ------------------------------------------: | -------: | ----------: | -----------------------------------: |
+| shp-source-default-commit-sha     |  `shp-source-${sourceType}-${resultName}`   | yes     |  at runtime | `${sourceType}` is `default` as is the only supported git source today |
+
+#### Future State of Build Results for Sources
+
+Assuming the following Build definition:
+
+```yaml
+apiVersion: shipwright.io/v1alpha1
+kind: Build
+metadata:
+  name: a-build
+spec:
+  sources:
+  - name: a-git-source
+    type: git
+    url: https://github.com/shipwright-io/sample-go
+  - name: an-http-source
+    type: http
+    url: https://licenses.company.com/customer-id/license.tar
+```
+
+We will have the following:
+
+| Result Name                       | Naming Convention                           | Emitted Today | Definition  |
+| --------------------------------: | ------------------------------------------: |-------: | ----------: |
+| shp-source-a-git-source-commit-sha     |  `shp-source-${spec.sources[0].name}-${resultName}` | no     |  at runtime |
+| shp-source-a-git-source-commit-author     |  `shp-source-${spec.sources[0].name}-${resultName}` | no     |  at runtime |
+| shp-source-an-http-source-size     |  `shp-source-${spec.sources[1].name}-${resultName}` | no     |  at runtime |
+
+#### Build Results Sources API Proposal
+
+Because of the future adoption of sources, the proposed API for `Build Results` should consider the following:
+
+- The _.status_ subresource should provide a relationship between a source and the results emitted in it´s step definition.
+- The _.status_ subresource should be able to surface results emitted from different sources.
+
+Therefore the **proposal** is to introduce `_.status.sources_` subresource which holds a list of `SourcesResult`:
+
+```go
+type KeyValue map[string]string
+
+type Sources struct {
+  Results []SourceResult `json:"sources"`
+}
+
+type SourcesResult struct {
+  Name   string   `json:"name"`
+  Git    KeyValue `json:"git,omitempty"`
+  Http   KeyValue `json:"http,omitempty"` // A new field can be always introduced to support new types, e.g. bundle
```

Comment:
I am in favor of strongly typing this for three reasons:

1. We know which results we produce.
2. `kubectl explain buildrun.status.sources.git` will tell you that there is a commit-sha which is only possible if we explicitly define this.
3. We are not forced to use `string` as type. The size of an HTTP source for example would be numeric.

I also think that you need to make it a reference. At least I cannot imagine what `omitempty` would do for an object type (= what makes it empty).

Reply:
To clarify what I meant:

```go
type GitSourceResult struct {
  CommitSha string `json:"commit-sha"`
}

type HttpSourceResult struct {
  Size int `json:"size"`
}

type SourceResult struct {
  Name   string   `json:"name"`
  Git    *GitSourceResult  `json:"git"`
  Http   *HttpSourceResult `json:"http"` // A new field can be always introduced to support new types, e.g. bundle
}

type BuilRunStatus struct {
  Sources []SourceResult `json:"sources"`
}
```